syntax = "proto3";

import "google/protobuf/empty.proto";  // Import Empty type
/* 
Command to create stub files:
python -m grpc_tools.protoc --proto_path=. ./proto/cache_coordinator.proto --python_out=. --grpc_python_out=.
*/
service CacheCoordinatorService {
  //rpc RegisterJob(JobInfo) returns (RegisterJobResponse);
  //rpc SendMetrics(MetricsRequest) returns (google.protobuf.Empty);
  rpc Ping(PingRequest) returns (PingResponse);
  rpc RegisterJob(RegisterJobRequest) returns (RegisterJobResponse);
  rpc GetNextBatchToProcess(GetNextBatchRequest) returns (GetNextBatchResponse);
  rpc JobEnded(JobEndedRequest) returns (google.protobuf.Empty);
  rpc GetDatasetInfo(DatasetInfoRequest) returns (DatasetInfoResponse);

}


message DatasetInfoRequest {
  string data_dir = 2;
  }

message DatasetInfoResponse {
  int32 num_files = 1;
  int32 num_chunks = 2;
  int32 chunk_size = 3;

  }

message PingRequest {
  string message =1;
}

message PingResponse {
  string message =1;
}

message RegisterJobRequest {
  int32 job_id = 1;
  string data_dir = 2;
  }

message RegisterJobResponse {
  bool job_registered = 1;
  string message =2;
  }

  message GetNextBatchRequest {
  int32 job_id = 1;
  int32 num_batches_requested = 2;
  }

message GetNextBatchResponse {
  int32 job_id = 1;
  repeated Batch batches = 2;
  }

message Batch {
  string batch_id = 1;
  repeated int32 indicies = 2;
  bool is_cached = 3;
}

 message JobEndedRequest {
    int32 job_id = 1;
}


/*
message GetBatchStatusRequest {
  int64 batch_id = 1;
  string dataset_id =2;
}

message GetBatchStatusResponse {
  bool batch_cached_or_in_progress = 1;
  string message =2;
  }

 message JoEnded {
    int32 job_id = 1;
}

  message JobMetricsInfo {
    int32 job_id = 1;
    string dataset_id =2;
    string metrics = 3;
}



message RegisterDatasetResponse {
  bool dataset_registered = 1;
  string message =2;
  }
  message RegisterDatasetInfo {
  string dataset_id =1;
  string data_dir =2;
  string transformations =3;
  string labelled_samples = 4;
}

message BatchAccessPatternList {
    int64 job_id = 1;
    repeated Batch batches = 2;
    string dataset_id =3;
}

message Batch {
  int64 batch_id = 1;
  repeated int32 sample_indices = 2;
}

message Message {
  string message =1;
  }
  */
